### PR TITLE
TLS 1.3: Add Finished Message and wrapup

### LINKED ISSUE #5497: Upstream TLS 1.3 server Finished message writing
This issue tracks the upstreaming of the server-side Finished message writing routines in TLS 1.3.

Source: https://github.com/hannestschofenig/mbedtls/tree/tls13-prototype

The Finished message writing routines in ssl_tls13_generic.c have to be extended (if needed) to support the server side.

Precise acceptance criteria: tests with GnuTLS/OpenSSL clients where the TLS 1.3 Mbed TLS server sends a Finished message that is successfully parsed by the client.

### LINKED ISSUE #5498: Upstream TLS 1.3 client Finished message parsing
This issue tracks the upstreaming of the server-side Finished message parsing routines in TLS 1.3.

Source: https://github.com/hannestschofenig/mbedtls/tree/tls13-prototype

The Finished message parsing routines in ssl_tls13_generic.c have to be extended (if needed) to support the server side.

Precise acceptance criteria: tests with GnuTLS/OpenSSL clients where the TLS 1.3 Mbed TLS server receives a Finished message and successfully parsed it.

### PR DESCRIPTION
**Pull request by yuhaoth**
fixes #5497 
fixes #5498
preceding PR: #5748

### CODE REVIEW

File: library/ssl_tls13_server.c [hunk 1]
```diff
@@ -1476,6 +1476,63 @@ static int ssl_tls13_write_certificate_verify( mbedtls_ssl_context *ssl )
 }
 #endif /* MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED */
 
+/*
+ * State Handler: MBEDTLS_SSL_SERVER_FINISHED
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1480]:
```suggestion
 * Handler for MBEDTLS_SSL_SERVER_FINISHED
```

File: library/ssl_tls13_server.c [hunk 2]
```diff
@@ -1476,6 +1476,63 @@ static int ssl_tls13_write_certificate_verify( mbedtls_ssl_context *ssl )
...
+    ret = mbedtls_ssl_tls13_write_finished_message( ssl );
+    if( ret != 0 )
+        return( ret );
+
+    mbedtls_ssl_handshake_set_state( ssl, MBEDTLS_SSL_CLIENT_FINISHED );
+    return( 0 );
+}
+
+/*
+ * State Handler: MBEDTLS_SSL_CLIENT_FINISHED
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1495]:
```suggestion
 * Handler for MBEDTLS_SSL_CLIENT_FINISHED
```

File: library/ssl_tls13_server.c [hunk 3]
```diff
@@ -1476,6 +1476,63 @@ static int ssl_tls13_write_certificate_verify( mbedtls_ssl_context *ssl )
...
+    if( ret != 0 )
+        return( ret );
+
+    mbedtls_ssl_handshake_set_state( ssl, MBEDTLS_SSL_HANDSHAKE_WRAPUP );
+    return( 0 );
+
+}
+
+/*
+ * State Handler: MBEDTLS_SSL_FLUSH_BUFFERS
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1513]:
```suggestion
 * Handler for MBEDTLS_SSL_FLUSH_BUFFERS
```

File: library/ssl_tls13_server.c [hunk 4]
```diff
@@ -1476,6 +1476,63 @@ static int ssl_tls13_write_certificate_verify( mbedtls_ssl_context *ssl )
...
+    MBEDTLS_SSL_DEBUG_MSG( 1,
+                  ( "Switch to handshake traffic keys for outbound traffic" ) );
+    mbedtls_ssl_set_inbound_transform( ssl, ssl->handshake->transform_handshake );
+    ret = mbedtls_ssl_tls13_process_finished_message( ssl );
+    if( ret != 0 )
+        return( ret );
+
+    mbedtls_ssl_handshake_set_state( ssl, MBEDTLS_SSL_HANDSHAKE_WRAPUP );
+    return( 0 );
+
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1509]:
```suggestion
```

File: library/ssl_tls13_server.c [hunk 5]
```diff
@@ -1476,6 +1476,63 @@ static int ssl_tls13_write_certificate_verify( mbedtls_ssl_context *ssl )
...
+}
+
+/*
+ * State Handler: MBEDTLS_SSL_CLIENT_FINISHED
+ */
+static int ssl_tls13_process_client_finished( mbedtls_ssl_context *ssl )
+{
+    int ret;
+    MBEDTLS_SSL_DEBUG_MSG( 1,
+                  ( "Switch to handshake traffic keys for outbound traffic" ) );
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1501]:
```suggestion
                  ( "Switch to handshake traffic keys for inbound traffic" ) );
```

File: library/ssl_tls13_server.c [hunk 6]
```diff
@@ -1476,6 +1476,63 @@ static int ssl_tls13_write_certificate_verify( mbedtls_ssl_context *ssl )
...
+}
+
+/*
+ * State Handler: MBEDTLS_SSL_HANDSHAKE_WRAPUP
+ */
+static int ssl_tls13_handshake_wrapup( mbedtls_ssl_context *ssl )
+{
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "handshake: done" ) );
+
+    MBEDTLS_SSL_DEBUG_MSG( 1, ( "Switch to application keys for all traffic" ) );
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1528]: In client and server, we switch in the handshake wrap-up state to the application keys thus this code should probably be moved to ` mbedtls_ssl_tls13_handshake_wrapup`.

File: library/ssl_tls13_keys.h [hunk 7]
```diff
@@ -610,6 +610,19 @@ int mbedtls_ssl_tls13_key_schedule_stage_application( mbedtls_ssl_context *ssl )
...
+ *
+ * \param ssl  The SSL context to operate on. This must be in
+ *             key schedule stage \c Application, see
+ *             mbedtls_ssl_tls13_key_schedule_stage_application().
+ *
+ * \returns    \c 0 on success.
+ * \returns    A negative error code on failure.
+ */
+int mbedtls_ssl_tls13_generate_resumption_master_secret(
+    mbedtls_ssl_context* ssl );
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_keys.h:624]:
```suggestion
    mbedtls_ssl_context *ssl );
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

**Review by paul-elliott-arm — COMMENTED** [review]
Just to comment that I have reviewed up to this point, and have nothing more to add than Ronald did , also to point out that #5748 is now merged, so this could do with rebase.

**Comment by yuhaoth** [comment]
> Just to comment that I have reviewed up to this point, and have nothing more to add than Ronald did , also to point out that #5748 is now merged, so this could do with rebase.

Done. Rebased.

File: library/ssl_tls13_generic.c [hunk 8]
```diff
@@ -1310,6 +1383,13 @@ void mbedtls_ssl_tls13_handshake_wrapup( mbedtls_ssl_context *ssl )
 
     MBEDTLS_SSL_DEBUG_MSG( 3, ( "=> handshake wrapup" ) );
 
+    MBEDTLS_SSL_DEBUG_MSG( 1, ( "Switch to application keys for inbound traffic" ) );
+    mbedtls_ssl_set_inbound_transform ( ssl, ssl->transform_application );
+
+    MBEDTLS_SSL_DEBUG_MSG( 1, ( "Switch to application keys for outbound traffic" ) );
+    mbedtls_ssl_set_outbound_transform( ssl, ssl->transform_application );
+
+
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_generic.c:1392]:
```suggestion
```

File: library/ssl_tls13_generic.c [hunk 9]
```diff
@@ -1324,8 +1341,66 @@ static int ssl_tls13_prepare_finished_message( mbedtls_ssl_context *ssl )
...
+        }
+
+    }
+    else
+#endif /* MBEDTLS_SSL_CLI_C */
+
+#if defined(MBEDTLS_SSL_SRV_C)
+    if( ssl->conf->endpoint == MBEDTLS_SSL_IS_SERVER )
+    {
+        mbedtls_ssl_key_set traffic_keys;
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply library/ssl_tls13_generic.c:1288]: Not addressed it seems.

File: library/ssl_tls13_keys.c [hunk 10]
```diff
@@ -1565,4 +1565,13 @@ int mbedtls_ssl_tls13_compute_handshake_transform( mbedtls_ssl_context *ssl )
     return( ret );
 }
 
+int mbedtls_ssl_tls13_generate_resumption_master_secret(
+    mbedtls_ssl_context *ssl )
+{
+    /* Erase master secrets */
+    mbedtls_platform_zeroize( &ssl->handshake->tls13_master_secrets,
```
[the remark this answers is not in the document]
Comment (ronald-cron-arm) [reply library/ssl_tls13_keys.c:1572]: Not addressed it seems.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]

**Comment by yuhaoth** [comment]
Open CI pass and internal CI still running

**Review by ronald-cron-arm — APPROVED** [review]
Thanks for the changes. LGTM.

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Comment by paul-elliott-arm** [comment]
Not happy to merge this without explanation for the CI fail. Its not a usual failure, and it does genuinely look like an issue.

**Comment by yuhaoth** [comment]
> Not happy to merge this without explanation for the CI fail. Its not a usual failure, and it does genuinely look like an issue.

OpenCI report same error again. I will try fix it .

**Comment by yuhaoth** [comment]
Update to fix CI fail. 
I can reproduce it in my local test. And it pass now. 
And I am not sure why internal CI report PASS for `test_no_x509_info`

**Review by paul-elliott-arm — APPROVED** [review]
LGTM (CI Pending)

**Comment by yuhaoth** [comment]
The openssl 1.1.1a exit with different exit code . 
In my local test , it return 0 and in OpenCI it return 1. 

with last commit, I expect internal CI will fail

File: tests/ssl-opt.sh [hunk 11]
```diff
@@ -11217,8 +11217,8 @@ requires_config_enabled MBEDTLS_SSL_SRV_C
 requires_openssl_tls1_3
 run_test    "TLS 1.3: Server side check - openssl with client authentication" \
             "$P_SRV debug_level=4 auth_mode=required crt_file=data_files/server5.crt key_file=data_files/server5.key force_version=tls13 tickets=0" \
-            "$O_NEXT_CLI -msg -debug -cert data_files/server5.crt -key data_files/server5.key -tls1_3" \
-            0 \
+            "$O_NEXT_CLI -msg -debug -cert data_files/server5.crt -key data_files/server5.key -tls1_3 -no_middlebox" \
+            1 \
```
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh:11221]: This is correct, client authentication cannot succeed as the support is not added yet (added in #5777).

**Comment by yuhaoth** [comment]
I skip the test. And It should be re-enabled in client auth task

**Review by ronald-cron-arm — APPROVED** [review]
I looked at the Internal CI not failing the "TLS 1.3: Server side check - openssl with client authentication" ssl-opt.sh test. The Openssl client does receive (in one chunk) the fatal alert message sent by the Mbed TLS server (unexpected message) but returns with zero anyway. When it is failing on the Open CI (as expected) the only difference I see is that the Openssl client receives the fatal alert message in two pieces instead of one piece and returns in error in that case. Thus all together I think the issue we encountered on the CI is rather an issue of error detection in ssl-opt.sh (the server side is not taken into account to determine the result of the test) and/or interaction between the Openssl client and ssl-opt.sh. Thus in my opinion nothing to prevent this PR to be merged with the aforementioned test being disabled. As @yuhaoth mentioned it will be re-enabled in #5777. @paul-elliott-arm what do you think about this ?

**Review by paul-elliott-arm — APPROVED** [review]
LGTM

**Pull request merged**